# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [ ]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [ ]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [ ]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [ ]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [ ]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [ ]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [ ]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [ ]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [ ]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [ ]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [ ]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [ ]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [ ]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [ ]:
data.duplicated().sum()

71

In [ ]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [ ]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [ ]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [ ]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [ ]:
#определим функцию с аргументом по столбцу
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [ ]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)
data

children  days_employed  dob_years education  education_id  \
0             1    8437.673028         42    высшее             0   
1             1    4024.803754         36   среднее             1   
2             0    5623.422610         33   среднее             1   
3             3    4124.747207         32   среднее             1   
4             0  340266.072047         53   среднее             1   
...         ...            ...        ...       ...           ...   
21520         1    4529.316663         43   среднее             1   
21521         0  343937.404131         67   среднее             1   
21522         1    2113.346888         38   среднее             1   
21523         3    3112.481705         38   среднее             1   
21524         2    1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

       total_income                       purpose total_income_category  \
0            253875                 покупка жилья                     B   
1            112080       приобретение автомобиля                     C   
2            145885                 покупка жилья                     C   
3            267628    дополнительное образование                     B   
4            158616               сыграть свадьбу                     C   
...             ...                           ...                   ...   
21520        224791             операции с жильем                     B   
21521        155999          сделка с автомобилем                     C   
21522         89672                  недвижимость                     C   
21523        244093  на покупку своего автомобиля                     B   
21524         82047         на покупку автомобиля                     C   

               purpose_category  
0      операции с недвижимостью  
1        операции с автомобилем  
2      операции с недвижимостью  
3         получение образования  
4            проведение свадьбы  
...                         ...  
21520  операции с недвижимостью  
21521    операции с автомобилем  
21522  операции с недвижимостью  
21523    операции с автомобилем  
21524    операции с автомобилем  

[21331 rows x 14 columns]

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [ ]:
# определим функцию для подсчета процентного соотношения, что сделает вывод в нашем исследовании более понятным 
def percentage(pdSerises):
    return str(round((pdSerises.sum() / pdSerises.count()) * 100, 2)) + '%'

# для обобщения данных и их наглядного представления построим сводную таблицу
data_pivot = data.pivot_table(index=['children'], values=["debt"], aggfunc=['sum', 'count', percentage])

# отсортируем значения по количеству детей , чтобы получить упорядоченную картину
data_pivot = data_pivot.sort_values(by=('children'), ascending=True)
data_pivot

sum  count percentage
          debt   debt       debt
children                        
0         1063  14091      7.54%
1          444   4808      9.23%
2          194   2052      9.45%
3           27    330      8.18%
4            4     41      9.76%
5            0      9       0.0%

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [ ]:
# для обобщения данных и их наглядного представления построим сводную таблицу
data_pivot = data.pivot_table(index=['family_status'], values=["debt"], aggfunc=['sum', 'count', percentage])

# отсортируем значения по процентному соотношению, чтобы получить упорядоченную картину
data_pivot = data_pivot.sort_values(by=('percentage', 'debt'), ascending=False)
data_pivot

sum  count percentage
                      debt   debt       debt
family_status                               
Не женат / не замужем  273   2796      9.76%
гражданский брак       385   4134      9.31%
женат / замужем        927  12261      7.56%
в разводе               84   1189      7.06%
вдовец / вдова          63    951      6.62%

**Вывод:** 

Исходя из результатов следует, если клиент когда-либо был в браке (категории `женат/замужем`, `в разводе`, `вдовец/вдова`), то вероятность задолженности будет ниже на 2-3 %, чем у заёмщика, который не состоит в официальном браке. Люди, состоящие в браке, должны принимать все решения совместно. Это касается и оформления банковской ссуды. Обычно в браке оба человека работают, поэтому выполнять взятые обязательства перед финансовой организацией, значительно легче. Можно предположить, что семейные отношения приучают к ответственности, в том числе и финансовой.

Несмотря на то, что в пункте 2.7.2 был сделан вывод о том, что семьи без детей имеют наименьшую задолженность, теперь можно предположить, для того чтобы являться надежным заёмщиком для банка, есть влияние нескольких критериев сразу - отсутствие детей и официальный брак.

Наименьшая доля просрочек по кредиту 6.62% выявлена у овдовевших супругов.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [ ]:
# # для обобщения данных и их наглядного представления построим сводную таблицу
data_pivot = data.pivot_table(index=['total_income_category'], values=["debt"], aggfunc=['sum', 'count', percentage])

# отсортируем значения по процентному соотношению, чтобы получить упорядоченную картину
data_pivot = data_pivot.sort_values(by=('percentage', 'debt'), ascending=False)
data_pivot

sum  count percentage
                       debt   debt       debt
total_income_category                        
E                         2     22      9.09%
C                      1353  15921       8.5%
A                         2     25       8.0%
B                       354   5014      7.06%
D                        21    349      6.02%

**Вывод:** 

Исходя из полученых данных в сводной таблице замечено, что в категориях доходов `A`, `D`, `E` количество данных намного меньше чем в остальных категориях. Тем самым мы столкнулись с особенностью данных, называемой выбросом (data outliers) или шумом в данных, которая, например, могла быть вызвана изначальными пропусками в столбце total_income. Поэтому данные в категориях доходов `A`, `D`, `E` не будут являться показателем надежности потенциального заёмщика в данном анализе. Для проведения достоверного анализа будет необходимо воспользоваться квантилями.

Определим зависимость исходя из результатов по категориям `B`, `C`. Процент просрочки кредита немного меньше у заемщиков  категории `B`, с доходом 200,001–1,000,000 рублей (невозврат составляет 7.06%), по-сравнению с категорией `C`, в которой процент невозврата кредита в срок составляет 8.5%. Исходя из этого можно вывести зависимость, чем выше уровень дохода, тем меньше вероятность просрочки. В то же время, вывод по категориям `B` и `C` можно распространить на все группы, так как суммарно в группах `B` и `C` находятся 98% клиентов.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [ ]:
# построим сводную таблицу для ответа на вопрос
data_pivot = data.pivot_table(index=['purpose_category'], values=["debt"], aggfunc=['sum', 'count', percentage])

# отсортируем значения по процентному соотношению, чтобы получить упорядоченную картину
data_pivot = data_pivot.sort_values(by=('percentage', 'debt'), ascending=False)
data_pivot

sum  count percentage
                         debt   debt       debt
purpose_category                               
операции с автомобилем    400   4279      9.35%
получение образования     369   3988      9.25%
проведение свадьбы        183   2313      7.91%
операции с недвижимостью  780  10751      7.26%

**Вывод:** 

В полученной сводной таблице нет сильно отличающихся по количеству категорий, тем самым результат можно интерпретировать следующим образом: 
- заёмщики, у которых целью кредита являются `операции с автомобилем`(процент просрочек 9.35%) или `получение образования`(процент просрочек 9.25%) более склонны к невозврату кредита в срок. 
- категории клиентов, которые берут кредиты на `проведение свадьбы` (процент просрочек 7.91%) и `операции с недвижимостью` (процент просрочек 7.26%) менее склонны к невозврату кредита в срок.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Традиционными причинами, приводящими к появлению пропусков, могут быть: невозможность получения данных или их обработки; искажение или сокрытие информации; фрод; технические причины.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* В случае когда некоторые значения данных сильно выделяются среди большинства других, использование среднего значения может некорректно характеризовать данные. Медиана лишена недостатков среднего значения, так как на нее не влияют выбросы или "шумы" в данных. 

### Шаг 4: общий вывод.

При проведении исследования данных, были выявлены зависимости между семейным положением клиента и количеством детей на факт погашения кредита в срок. Максимальный процент невозврата кредита у клиентов не состоящих в официальном браке, а также у клиентов имеющих детей. Клиенты, входящие в данные категории являются клиентами с максимальной степенью риска невозврата кредита в положенный срок. 

Мы также видим, чем выше уровень дохода, тем меньше вероятность просрочки по возврату кредита. Несмотря на то, что особенность данных в категории уровня дохода заемщиков не позволила нам дать исчерпывающий ответ по всем категориям, вывод по категориям `B` и `C` можно распространить на все остальные категории, так как суммарно в группах В и С около 98% клиентов.
    
Исходя из проведенного нами анализа, можно спроектировать профиль наиболее платежеспособного клиента следующим образом:
    
*Человек состоящий в браке, без наличия детей, с высоким уровнем дохода, и целью получения кредита которого являются операции связанные с недвижимостью.*